Resolves issue #114: Update mod doc repo example and instructions (#120)
redhat-documentation/modular-docs: modular-docs-manual/content/topics/module_modular-documentation-repositories.adoc

[id="modular-documentation-repositories"]
= Modular Documentation Repositories

> IMPORTANT: As CCS strategic imperatives and Pantheon 2 development continue to evolve, the details and recommendations in this section and possibly other sections in this guide might be modified. The Flexible Customer Content (FCC) team recognizes that the fluctuation in recommendations can be frustrating for CCS teams trying to adopt modularization standards and apologizes in advance as the initiative pursues the best possible outcome for CCS documentation.

Reviewer: Again... internal RH reference. See https://github.com/redhat-documentation/modular-docs/pull/121#discussion_r449559189

> Due to many factors that determine documentation repository design for individual Red Hat products, the Flexible Customer Content (FCC) initiative cannot provide a strict documentation repository template for all CCS teams to follow. Each team faces different implementation challenges, especially when Red Hat product documentation is single-sourced with corresponding community project documentation.

Reviewer: I suggest rewording to avoid mentions of FCC, CCS, etc. This paragraph is applicable to upstream and other organizations as well.

> However, for optimal FCC, Pantheon, and Pantheon 2 compliance, the FCC initiative recommends a general repository structure similar to the following example, where possible:

Reviewer: I suggest rewording to avoid mentions of FCC, CCS, Pantheon, etc. This paragraph is applicable to upstream and other organizations as well.

Reviewer: In fact, I'd remove this paragraph altogether. This is a note that could be added to every single recommendation in these guidelines. As such, it serves no purpose here.

Author: I've added the note at the top of the entire doc, in the intro, but it started here because of emphasis. This issue has been particularly problematic with repo recommendations, so I just wanted to resurface it for clarity and great visibility. Regarding internal references, responding to that separately.

> .Recommended documentation repository structure (where possible)

Reviewer: I think this example should not copy an existing repo. Actual assembly and module names (instead of generics, such as `proc_procedure-module.adoc`) invite confusion.

Author: I've received overwhelming feedback for exactly the opposite in my mod doc repo trainings and other conversations with various teams. Having concrete, realistic, and various examples was and has been a huge ask. I got a lot of complaints for the former generic and simplified repo example.

[source]
----
├── _artifacts
├── _images
├── assemblies

>     ├── _artifacts -> ../_artifacts  // Symlinks can be extracted.

Reviewer: What does 'extracted' mean in this context?

Author: I explain in the bulletted description that follows. Can do so with variables, build tooling, scripting, whatever. I was trying to avoid a long callout here, and thus provided more detail in the description that follows.

    ├── _images -> ../_images
    ├── modules -> ../modules

>     ├── assembly-kogito-configuring.adoc

Reviewer: Let's use the underscore to separate prefixes. It's not a hard requirements, and hyphens are OK, but underscores are more commonly used and more practical from the mod docs perspective. So, for examples, I think we should go with them. (Regardless of what is selected, all examples in the guidelines need to be consistent.)

Author: That would mean a very large change to the entire repo, which uses hyphens. We've decided as a team, because there's so much discord over this, that we officially can't really lean to one or the other anymore, and both are viable options. So I'd suggest keeping this in the example, since whichever route we go, we'll likewise have someone coming in to argue for the other.

    ├── assembly-kogito-decision-engine.adoc
    ├── assembly-kogito-developing-decision-services.adoc
    ├── assembly-kogito-developing-process-services.adoc
    ├── assembly-kogito-using-dmn.adoc
    ├── assembly-kogito-using-drl.adoc
    ├── ...
├── modules
    ├── configuration
        ├── con-data-index-service.adoc
        ├── con-jobs-service.adoc
        ├── proc-data-index-service-enable-security.adoc
        ├── ref-data-index-service-queries.adoc
        ├── ...
    ├── creating-running
        ├── con-kogito-automation.adoc
        ├── con-kogito-modelers.adoc
        ├── proc-kogito-app-examples-running.adoc
        ├── ref-kogito-app-examples.adoc
        ├── ...
    ├── decision-services  // Contains modules related to multiple assemblies.
        ├── con-decision-tables.adoc
        ├── con-dmn.adoc
        ├── con-drl.adoc
        ├── proc-dmn-model-creating.adoc
        ├── ref-dmn-model-example.adoc
        ├── ...
    ├── process-services
        ├── ...
    ├── ...
├── titles-enterprise
    ├── kogito-configuring
        ├── _images -> ../../_images
        ├── assemblies -> ../../assemblies
        ├── master.adoc

>         └── master-docinfo.xml

Reviewer: Docinfo files are specific to ccutil and Red Hat. I don't think they should be included in this.

    ├── kogito-creating-running
        ├── ...
    ├── kogito-developing-decision-services  // Contains multiple assemblies.
        ├── ...
    ├── ...

> ├── pantheon2.yml

Reviewer: Specific to RH.

└── README.md
----

> This repository structure contains the following key components:

Reviewer: I'd suggest using callouts for these descriptions.

Author: Haha, funny you say that. There has been explicit guidance from the style team to avoid callouts where possible, for accessibility reasons and for minimizing clutter. So I'm afraid I'm stuck between two conflicting opinions on this.

* `_artifacts`: Directory with document attribute files and other metadata files. The underscore in the folder name is an optional indication of a non-content folder.
* `_images`: Directory with image files, optionally divided into sub-directories by subject matter. The underscore in the folder name is an optional indication of a non-content folder.
* `assemblies`: Directory with assembly files. You collect and compile assemblies into documents (titles) using the `titles-enterprise` folder.
* `modules`: Directory with module files, optionally divided into sub-directories by subject matter.

> * `titles-enterprise`: Directory with folders for each enterprise document or _title_ to be built in Pantheon. Each document folder contains a `master.adoc` file where you include one or more relevant assemblies that are built as part of the document. The `master-docinfo.xml` file contains metadata that is required by `bccutil` (and `ccutil`) when converting to DocBook format. (This metadata file might be deprecated in a future release of `bccutil` and Pantheon 2.) In repositories that are single-sourced with a community project, a separate `titles-community` folder might be dedicated to community configurations and output, if applicable.

Reviewer: Let's not refer to internal RH tooling.

> * `pantheon2.yml`: YAML file with metadata that is required for publishing content on Pantheon 2:

Reviewer 2: No need to say "Directory"/"YAML file" imo. That should be obvs to users. Start with third person singular verbs.

* `_artifacts`: Contains attributes and metadata. Optionally prefix the name with an underscore character to denote a directory that does not contain content.
* `_images`: Contains image files that you can optionally organize into sub-directories. ... and so on.

"In cases where content exists in community repositories, you can create a `titles-community` directory that holds upstream only content."

Author: I'm listing things that are all at the same top level of the repo, most of which directories, some files (like the YAML and README). There's really not a directory to specify since it depends on the user and not shown in the structure. I could just use an example parent dir name, but then should add the same to all the other directories listed for consistency (example/titles-enterprise). Then just noise to me. Seems unnecessary imo.

Yeah, would be fine to role with verb-first, but because some are folders and the one is a file, I wanted to start with what it is (directory or file), otherwise when I get to the YAML, it will also say "Contains ..." and could be odd, even with the parent dire there named. Not trying to split hairs, just keep it simple and clear. I think people get it, but that's just me. I'm open if you have a strong feeling otherwise.

Regarding that community rewording, I don't think that's quite accurate. The whole purpose of a single-sourced repo is to have mostly or all shared content, and then the community folder dedicated to the configs to just the community rendering, which is really all that's different. For content that's community only, that's controlled by including it in the community folder rendering files and not including in the enterprise rendering files. But again, this may vary depending on circumstances, so the description I went with is for this specific repo scenario, which would follow what I've explained here.

Reviewer 2: YAML files can contain metadata.

Sure thing with that community rewording. I just see passive and my instinct is to re-word.

Author: Understood.

> * `pantheon2.yml`: YAML file with metadata that is required for publishing content on Pantheon 2:

Reviewer: This is RH internal (unfortunately, despite promises to the contrary, Pantheon is not being developed as an open(-source) project. Therefore, references to the tool should be kept out of these guidelines, so as not to detract from their value for other organizations and upstream projects.

If we need internal RH guidelines for publishing, they should be kept separately from this content.

+
.Example `pantheon2.yml` file
[source,yaml]
----
server: http://localhost:8080
repository: mod-doc-repo-example
variants:
   - name: global-attributes
     path: _artifacts/document-attributes.adoc
     canonical: true
   - name: community-attributes
     path: _artifacts/document-attributes-kogito-comm.adoc
   - name: enterprise-attributes
     path: _artifacts/document-attributes-kogito-ent.adoc
assemblies:
  - assemblies/*.adoc
modules:
  - modules/**/*.adoc
resources:

>   - _images/**/*.png

Author: Who/where would we need to go to for verification? Would love to confirm as part of this PR.

>   - _images/**/*.png

Reviewer 3: Well I just tested it and it appears that it is not supported, unfortunately.

Assuming that this is a feature that CCS would want to have, would you mind creating a Jira ticket for it and then we can work with Dawn to prioritize it?

>   - _images/**/*.png

Author: Per gchat with @benradey: Updated PR for now to be accurate and created CCSOPS ticket to request support for `/**/` notation for any number of sub-folders.

  - _artifacts/*.adoc
----
+
For information about configuring this file for your specific content, contact a member of the CCS tooling team.

This repository structure offers the following potential advantages:

> * *Flexibility in determining publishable titles.* In this structure, the images, artifacts, assemblies, and modules all serve as distinct building blocks that you use to compile a document or _title_. You can include an image in multiple modules, a module in multiple assemblies, and an assembly in multiple titles. Additionally, just as an assembly can include multiple modules, a title can contain multiple assemblies.

Reviewer 2: ```suggestion
* *Flexibility in determining publishable titles.* In this structure, the images, artifacts, assemblies, and modules all serve as distinct building blocks that you use to compile a document or _title_. You can include an image in multiple modules, a module in multiple assemblies, and an assembly in multiple titles. Additionally, just as assemblies can include multiple modules, titles can contain multiple assemblies.
```

Author: Plural "assemblies" containing multiple modules and "titles" containing multiple assemblies doesn't give the emphasis of multiple in the same assembly/title that we're trying to convey. Two titles each with one assembly still = plural titles with multiple assemblies. I was trying to emphasize multiple in one.

Boy, I'm sure being stiff on this PR so far but really not trying to :( I can give up and go with the flow if I need to. :)

Reviewer 2: 2 in 1 shampoo never made sense to me. 1 isn't big enough to hold 2. That's why there's 2.

Jokes aside, it's fine if you don't want to make the change. I generally prefer using plural forms for consistency. Some day I'll have to explain the reason behind that but I can't put it into words at the moment.

Author: Haha! Good one :)  Well I'll look forward to that rationale (honestly, not being sarcastic at all). Sounds interesting. For now, will leave as is, if okay with you.

* *More intuitive content reuse.* When you decouple the standalone content components from the publishable titles, you avoid potentially awkward includes when modules or assemblies from one title are reused in another. For example, consider a repository where each publishable title folder contains all modules or assemblies or both directly within it. This creates a dependency for that module or assembly content on that title. As you reuse content across title folders, you end up with includes that direct to various other unrelated titles to retrieve the content. Alternatively, in the recommended repository structure, you can reuse all content independent of the titles in which you include it.

> * *Potentially easier collaboration in community and between content repositories.* For repositories that are single sourced with community projects or that need to reuse content between product documentation repositories, this structure enables you to isolate the enterprise title compilation requirements so that the standalone components can be shared more readily. However, single sourcing between community and enterprise content repositories poses other challenges beyond the scope of this document. (See the disadvantages and community considerations sections that follow.)

Reviewer 2: s/single sourcing/single-sourcing

Author: Yeah, I've also said "single-sourcing", but a growing counter argument is that it's only hyphenated because it's a clipped form of the full phrase "single-sourced publishing/content/whatever". But you're right, definitely becoming the norm and very logical. Changed throughout.

* *Consistency with Pantheon 2 YAML configurations.* In this structure, the division of modules, assemblies, and other resources is consistent with the division of components in the `pantheon2.yml` file that is required for documents published on Pantheon 2.

> This repository structure poses the following potential disadvantages:

Reviewer 2: 5 counts of the word "potential" here. I'd consider revising to drop/reword some of those instances to avoid overuse.

Author: Yeah, guilty :( I noticed it too, but I'm just so hesitant to be objective in some of these because of how many factors are at play. I ditched it from the two intro sentences before the lists. Left the other couple.

> * *Extensive use of symlinks.* In order to enable titles to be built with distinct assemblies, modules, and images, this repository structure uses many symlinks (symbolic links) for access across directories. You can extract these symlinks with a build script or additional variable configurations (such as `imagesdir`), but because this example repository is intended for a general CCS audience, it maintains explicit symlinks so that a user can build the repository titles with standard CCS and Asciidoctor tooling. Alternative repository structures that combine all relevant images, modules, and assemblies together in title folders can reduce the number of symlinks, but these alternative structures then forego many of the advantages listed previously.

Reviewer: ```suggestion
* *Extensive use of symlinks.* In order to enable titles to be built with distinct assemblies, modules, and images, this repository structure uses many symlinks (symbolic links) for access across directories. You can eliminate these symlinks with a build script or additional variable configurations (such as `imagesdir`), but because this example repository is intended for a general audience, it maintains explicit symlinks, so that a user can build the repository titles with standard Asciidoctor tooling. Alternative repository structures that combine all relevant images, modules, and assemblies together in title folders can reduce the number of symlinks, but these alternative structures then forego many of the advantages listed previously.
```

> * *Extensive use of symlinks.* In order to enable titles to be built with distinct assemblies, modules, and images, this repository structure uses many symlinks (symbolic links) for access across directories. You can extract these symlinks with a build script or additional variable configurations (such as `imagesdir`), but because this example repository is intended for a general CCS audience, it maintains explicit symlinks so that a user can build the repository titles with standard CCS and Asciidoctor tooling. Alternative repository structures that combine all relevant images, modules, and assemblies together in title folders can reduce the number of symlinks, but these alternative structures then forego many of the advantages listed previously.

Reviewer 2: While we're talking about symlinks, is it worth mentioning that Pantheon does not support `../` and requires symlink instead? 

It might be overkill for this doc but in some community projects symlinks are verboten. You can skirt around it with variables. 

For example, consider an includes directive like this: `includes::{modules}/proc_doing_something.adoc` 

In the title files you can then have one of the following:

Product/enterprise: `:modules: modules` (where `modules` is the name of the symlink)
Community: `:modules: ../path/to/modules`

Author: Definitely. @benradey also brought it up. His suggestion was for a section dedicated to considerations with symlinks. I agree with that but not up for it as part of this PR :)

Reviewer 2: Sounds good.

> * *Extensive use of symlinks.* In order to enable titles to be built with distinct assemblies, modules, and images, this repository structure uses many symlinks (symbolic links) for access across directories. You can extract these symlinks with a build script or additional variable configurations (such as `imagesdir`), but because this example repository is intended for a general CCS audience, it maintains explicit symlinks so that a user can build the repository titles with standard CCS and Asciidoctor tooling. Alternative repository structures that combine all relevant images, modules, and assemblies together in title folders can reduce the number of symlinks, but these alternative structures then forego many of the advantages listed previously.

Author: Good point @benradey. There's also another camp that is anti-symlinks as a whole, because they don't always work the same on every OS. Among us on CCS it's fairly safe, but on a community scale for example, most devs won't have it. So I'm hesitant to go into it here, but agree that you're right, a dedicated section with some guidance on symlinks would be good. I'll think on it and when I have some ideas, I'll create an issue for it.

> * *Potential conflicts with community content strategy.* Although this documentation repository structure is typically more collaboration-friendly than other FCC repository alternatives (as described in the previous advantages), the FCC strategy as a whole still might not be well received among all community projects. For example, many community projects prefer Mavenized repositories divided into chapter files instead of module files, all categorized in folders by guides or pages (titles) to be built. A community aversion to modularization in general can require writers to modify this example repository design, to pursue tooling alternatives to accommodate all content needs, or in some cases to abandon single-sourcing with community.

Reviewer: Redundant references to (internal) FCC.

The whole paragraph is a speculation. I don't think it needs to be here.

Author: Right, understood. I explained to the group in the call about this that a recurring complaint I keep receiving in my courses and work with other teams is that the mod doc ref guide is great at explaining what to do, but not really why. Teams with differing strategies and opinions are eager for the rationale in some things, particularly with repo design. So in my initial research and planning for this repo example and documentation, one big request was to please explain the advantages and disadvantages of the general strategy here, since it's up to individual teams to assess and decided, based on their many factors. So some of this is indeed speculation and potential, but that's what I've been asked by many to provide. So I'd prefer to keep it as an answer to some of the potential why's.

To view this example documentation repository, download the `mod-doc-repo-example.zip` file from the https://github.com/redhat-documentation/modular-docs/tree/master/modular-docs-manual/files[modular documentation source repository].

To build the example titles, in a command terminal, navigate to `~/titles-enterprise/__DOC_FOLDER__` and enter one of the following commands:

.Build with `ccutil` (deprecated)
[source]
----
$ ccutil compile --lang=en-US --format html-single --open

> ----

Reviewer 2: If ccutil is deprecated, let's not mention it if this is for newbies.

Reviewer 2: Or maybe move ccutil below bccutil and asciidoctor.

Author: You're right. I shouldn't say "deprecated". I was using the tooling language and they consider it deprecated. It just applies to RHEL7 and earlier. I still use it for example, on RHEL 7 CSB. I've reworded to just say "(RHEL 7 or Fedora 29 and earlier)" to match the bccutil snippet. I swapped the ordering though as you said to go from most recent/desired to least. Good call.

.Build with `ccutil` (RHEL 7 or Fedora 29 and earlier)
[source]
----
$ ccutil compile --lang=en-US --format html-single --open

> ----

Reviewer: Internal references. This tooling is not available publicly.

[id="modular-documentation-repositories"]
= Modular Documentation Repositories

> Due to many factors that determine documentation repository design for individual products, the modular documentation initiative cannot provide a strict documentation repository template for all documentation teams to follow. Each team faces different implementation challenges, especially when Red Hat enterprise product documentation is single-sourced with corresponding community project documentation.

Reviewer: Keep it simple?
```suggestion
Due to many factors that determine repository design for individual products, these modular documentation guidelines do not provide a strict documentation repository template.
```

Author: Normally I'd agree, but I want to avoid being so concise and vague here. I want it crystal clear why we can't provide the strict template. Just saying "due to many factors" alone isn't sufficient, imo. So I'd prefer to keep the wording as is.

Reviewer: I don't like "due to many factors", too, and wanted to cut even that...

My point is that it's a manual. Manuals should instruct while being as concise as possible -- not elaborate on reasons regarding the inclusion or exclusion of this or that. If you feel so strongly about it, I suggest we create an FAQ section at the end.

Author: Okay @rkratky , I give up :) For the sake of getting this through, I've taking your section here and below. Done.

> However, for a general repository structure that the modular documentation initiative recommends for optimal results with modular standards, see the example documentation repository and `README` file in the https://github.com/redhat-documentation/modular-docs/tree/mod-doc-repo-example[`mod-doc-repo-example`] branch in this repository.

Reviewer: ```suggestion
For a general repository structure used by documentation teams at Red Hat, see the example documentation repository and `README` file in the https://github.com/redhat-documentation/modular-docs/tree/mod-doc-repo-example[`mod-doc-repo-example`] branch in this repository.
```

Author: Again, wanting to be clear, but also, this rewording is necessarily correct. It's not a repo design "used by doc teams at Red Hat", at least not yet. Only a few teams are somewhat similar to this. Otherwise, teams are all over the map. That's kind of the purpose of the repo example, to give a recommended structure for optimal results with mod doc. So I guess I again would prefer the existing wording over the rewording.

Reviewer: * I want to make it clear that the example is RH-specific -- as discussed. We can qualify the statement ("some teams...").
* I I think it's superfluous to include any wording referring to a "modular docs initiative". This entire manual is a product of some initiative. In product docs, we don't say "The Customer Content Services team recommends..." :)
```suggestion
For a general repository structure used by some documentation teams at Red Hat, see the example documentation repository and `README` file in the https://github.com/redhat-documentation/modular-docs/tree/mod-doc-repo-example[`mod-doc-repo-example`] branch in this repository.
```
